src/core/dyad_core.h

```diff
@@ -0,0 +1,192 @@
+    bool intercept; // if true, intercepts (f)open and (f)close
+};
+extern const struct dyad_ctx dyad_ctx_default;
+typedef struct dyad_ctx dyad_ctx_t;
+
+struct dyad_kvs_response {
+    char *fpath;
+    uint32_t owner_rank;
+};
+typedef struct dyad_kvs_response dyad_kvs_response_t;
+
+// Debug message
+#ifndef DPRINTF
+#define DPRINTF(curr_dyad_ctx, fmt,...) do { \
+    if (curr_dyad_ctx && curr_dyad_ctx->debug) fprintf (stderr, fmt, ##__VA_ARGS__); \
+} while (0)
+#endif // DPRINTF
+
+#define TIME_DIFF(Tstart, Tend ) \
+    ((double) (1000000000L * ((Tend).tv_sec  - (Tstart).tv_sec) + \
+                              (Tend).tv_nsec - (Tstart).tv_nsec) / 1000000000L)
+
+// Detailed information message that can be omitted
+#if DYAD_FULL_DEBUG
+#define IPRINTF DPRINTF
+#define IPRINTF_DEFINED
+#else
+#define IPRINTF(curr_dyad_ctx, fmt,...)
+#endif // DYAD_FULL_DEBUG
+
+
+/**
+ * @brief Intialize the DYAD context
+ * @param[in]  debug          true if debugging is enabled
+ * @param[in]  check          ???
+ * @param[in]  shared_storage indicate if the storage associated
+ *                            with the path is shared
+ * @param[in]  key_depth      depth of the key hierarchy for the path
+ * @param[in]  key_bins       number of bins used in key hashing
+ * @param[in]  kvs_namespace  Flux KVS namespace to be used for this
+ *                            instance of DYAD
+ * @param[out] ctx            the newly initialized context
+ *
+ * @return An integer error code (values TBD)
+ */
+int dyad_init(bool debug, bool check, bool shared_storage,
+        unsigned int key_depth, unsigned int key_bins,
+        const char *kvs_namespace, const char* prod_managed_path,
+        const char *cons_managed_path, bool intercept, dyad_ctx_t **ctx);
+
+/**
+ * @brief Wrapper function that performs all the common tasks needed
+ *        of a producer
+ * @param[in] ctx    the DYAD context for the operation
+ * @param[in] fname  the name of the file being "produced"
+ *
+ * @return An integer error code (values TBD)
+ */
+#if DYAD_PERFFLOW
+__attribute__((annotate("@critical_path()")))
+#endif
+int dyad_produce(dyad_ctx_t *ctx, const char *fname);
+
+/**
+ * @brief Wrapper function that performs all the common tasks needed
+ *        of a consumer
+ * @param[in] ctx    the DYAD context for the operation
+ * @param[in] fname  the name of the file being "consumed"
+ *
+ * @return An integer error code (values TBD)
+ */
+#if DYAD_PERFFLOW
+__attribute__((annotate("@critical_path()")))
+#endif
+int dyad_consume(dyad_ctx_t *ctx, const char *fname);
+
+/**
+ * @brief Commits (i.e., writes) the file's metadata to the Flux KVS
+ * @param[in] ctx    the DYAD context for the operation
+ * @param[in] fname  the name of the file being committed
+ *
+ * @return An integer error code (values TBD)
+ */
+#if DYAD_PERFFLOW
+__attribute__((annotate("@critical_path()")))
+#endif
+int dyad_commit(dyad_ctx_t *ctx, const char *fname);
```

Comment:
We should not have this function in this header. It should be moved to dyad_core.c and marked as static